### PR TITLE
Noncontextual sets

### PR DESCRIPTION
**Pull request by TimWeaving**
Sizeable change to ContextualSubspace - introduced the new StabilizeFirst noncontextual strategy which is conceptually different from the existing methods. This works on the premise that a set of stabilizers has already been identified (for example, through an auxiliary operator preserving scheme) and therefore we construct a noncontextual operator that respects these.

### CODE REVIEW

File: symmer/symplectic/noncontextual_op.py [hunk 1]
```diff
@@ -137,6 +145,21 @@ def _single_sweep_noncontextual_operator(cls, H, strategy='magnitude'):
         
         return cls.from_PauliwordOp(noncontextual_operator)
 
+    @classmethod
+    def _from_basis_noncontextual_op(cls, H: PauliwordOp, basis: PauliwordOp):
```
Comment (TimWeaving) [inline symmer/symplectic/noncontextual_op.py:149]: For use with the from_hamiltonian initialization method for NoncontextualOp - the users specifies a noncontextual generating set and a corresponding noncontextual Hamiltonian is built by taking any terms in H that may be constructed under the Jordan product.

File: symmer/projection/utils.py [hunk 2]
```diff
@@ -51,7 +51,7 @@ def update_eigenvalues(
     stabilizers.coeff_vec = (-1) ** np.count_nonzero(
         np.bitwise_and(
             reconstruction, 
-            basis.coeff_vec==-1
+            np.asarray(basis.coeff_vec)==-1
```
Comment (TimWeaving) [inline symmer/projection/utils.py:54]: We often find these sorts of bugs in Symmer, whereby something that _should_ be an array is not (maybe a list or tuple) and thus the behaviour is not as intended.

File: symmer/projection/contextual_subspace.py [hunk 3]
```diff
@@ -136,6 +185,31 @@ def _random_stabilizers(self,
         
         return S
 
+    def _get_clique_representatives(self, symmetry_terms=None, n_cliques=2, clique_reps = []):
```
Comment (TimWeaving) [inline symmer/projection/contextual_subspace.py:188]: This method identifies a list of anticommuting operators that commute with the selected stabilizers (the required noncontextual structure). It is also possible to force stabilizers within a clique by starting recursion with clique_reps non-empty.

File: symmer/projection/utils.py (hunk 2 above)
Comment (AlexisRalli) [reply symmer/projection/utils.py:54]: Mmm this is true. I think what you have done here isnt' the best though... We should probably make the variable an array when basis.coeff_vec is set... Do you agree? We could maybe go through the whole repo and check anytime we allow people to input lists.

File: symmer/projection/contextual_subspace.py [hunk 4]
```diff
@@ -14,19 +14,30 @@ def __init__(self,
             noncontextual_strategy: str = 'diag',
             unitary_partitioning_method: str = 'LCU',
             reference_state: np.array = None,
-            noncontextual_operator: NoncontextualOp = None
+            noncontextual_operator: NoncontextualOp = None,
         ):
         """
```
Comment (AlexisRalli) [inline symmer/projection/contextual_subspace.py:19]: We need doc strings 👼

File: symmer/symplectic/noncontextual_op.py (hunk 1 above)
Comment (AlexisRalli) [reply symmer/symplectic/noncontextual_op.py:149]: Nice!

Comment (AlexisRalli) [reply symmer/symplectic/noncontextual_op.py:149]: (we should include the Jordan product definition in our doc strings)

File: symmer/symplectic/noncontextual_op.py [hunk 5]
```diff
@@ -137,6 +145,21 @@ def _single_sweep_noncontextual_operator(cls, H, strategy='magnitude'):
...
+        """ Construct a noncontextual operator given a noncontextual basis.
```
Comment (AlexisRalli) [inline symmer/symplectic/noncontextual_op.py:150]:
```suggestion
        """ Construct a noncontextual operator given a noncontextual basis, via the Jordan product ( regular matrix product if the operators commute, and equal to zero if the operators anticommute.)
```

File: symmer/projection/contextual_subspace.py (hunk 4 above)
Comment (TimWeaving) [reply symmer/projection/contextual_subspace.py:33]: Done!

**Review by AlexisRalli — APPROVED** [review]

**Pull request merged**